feat: performance improvements

leonardoventurini opened the pull request:
OSS-604

Due to the nature of the iterative research, this PR ended up large, mixing plain refactor and file splits with actual improvements, I will point out the actual improvements to facilitate review.

In sum, this PR should greatly improve performance in Meteor 3, but requires further validation (WIP).

---

- Refactored AsynchronousQueue to support parallel processing with configurable batching
- Optimized MongoDB observers to prioritize initial document adds
- Added batched, concurrent task execution with optional ordered processing
- Performance improvements through parallel processing and better promise handling

https://github.com/meteor/meteor/pull/13445/commits/12993e9f2e8c53112b37c81f5bca7319d1745e1a

---

Improve compression defaults and include optimized version of `permessage-deflate` which supports setting a threshold of message size to start compressing.

This reduces CPU and memory significantly. Apps that consume large amounts of bandwidth but have CPU to spare can configure this via env var.

https://github.com/meteor/meteor/pull/13445/commits/b5fd7ac4308bec917d041bec3b75bab8b56dfa25

https://github.com/leonardoventurini/permessage-deflate2/pull/1

---

Added docs for `Performance > WebSocket Compression`:

https://github.com/meteor/meteor/pull/13445/commits/c73515470f5a4a623b8403231fbfa6c8472cd5b1

---

Other changes that didn't produce noticeable improvements, but worth keeping:

https://github.com/meteor/meteor/pull/13445/commits/748f7c4f29d87907af8eaeaa9e87d3c09fdef8c0
https://github.com/meteor/meteor/pull/13445/commits/4e8c7cf1815298da77f206c2b94d9b52f12e5e60
https://github.com/meteor/meteor/pull/13445/commits/6b5248e2367c061f0c8c34704f93166ca7f1c9d8

leonardoventurini committed:
simplify code

leonardoventurini committed:
adjust doc

leonardoventurini committed:
add test

leonardoventurini committed:
add setImmediate for inbound messages

leonardoventurini committed:
annotate memory growth point

leonardoventurini committed:
add test

leonardoventurini committed:
the result message is out of order

leonardoventurini committed:
refactor to oplog_v2_converter to typescript

leonardoventurini committed:
fix tests

leonardoventurini committed:
add docs back

leonardoventurini committed:
shrinkwrap

leonardoventurini committed:
refactor session and document views

leonardoventurini committed:
fix race condition in oplog tailing

leonardoventurini committed:
prevent blocking event loop in write fence

leonardoventurini committed:
rename files

leonardoventurini committed:
fix potential race condition dropping messages

leonardoventurini committed:
improve buffered writes flushing logic in order to not block the event loop

leonardoventurini committed:
extract stream handlers

leonardoventurini committed:
extract processors

leonardoventurini committed:
run adds in parallel

leonardoventurini committed:
send initial adds first; parallelize and batch asynchronous queue

leonardoventurini committed:
Merge remote-tracking branch 'origin/release-3.1.0' into leonardo/refactor-and-perf-phase-2

leonardoventurini committed:
fix check running twice

leonardoventurini committed:
ignore asynchronous queue from check

leonardoventurini committed:
Merge remote-tracking branch 'origin/release-3.1.0' into leonardo/refactor-and-perf-phase-2

leonardoventurini committed:
add timeout mechanism

leonardoventurini committed:
disable punycode warning on app process

leonardoventurini committed:
exclude roles from travis

leonardoventurini committed:
remove shrinkwraps from git (annoying things!)

leonardoventurini committed:
bump package versions so the sub deps are updated

leonardoventurini committed:
exclude roles in command

leonardoventurini committed:
update versions

leonardoventurini committed:
Merge branch 'release-3.1.0' into leonardo/refactor-and-perf-phase-2

leonardoventurini committed:
adjust compression parameters and use permessage-deflate2 with threshold (only messages larger than 1kb (512 chars) are compressed by default)

leonardoventurini committed:
Merge branch 'release-3.1.0' into leonardo/refactor-and-perf-phase-2

leonardoventurini commented:
Not sure if it is just flaky or an actual issue, adding it here:

<img width="1526" alt="image" src="https://github.com/user-attachments/assets/9fda6d2c-d237-4506-9176-d4ef477c7086">

alisnic commented:
This is very exciting 🤩

Grubba27 commented on packages/ddp-client/client/client_convenience.js:
```diff
@@ -1,6 +1,6 @@
 import { DDP } from '../common/namespace.js';
 import { Meteor } from 'meteor/meteor';
-import { loadAsyncStubHelpers } from "./queueStubsHelpers";
+import { loadAsyncStubHelpers } from "./queue_stub_helpers";
```
loved the name change!

Grubba27 commented on packages/meteor/asynchronous_queue.js:
This could be an easy win to typescript.


We only have 8 usages in our codebase

leonardoventurini replied:
We can't use TypeScript inside this package easily, not entirely sure why

Grubba27 commented on packages/mongo/observe_handle.ts:
```diff
@@ -15,11 +15,14 @@ export class ObserveHandle {
+  public initialAddsSentResolver: (value: void) => void = () => {};
+  public initialAddsSent: Promise<void>
+
+  _added?: (...args: any[]) => Promise<void>;
+  _addedBefore?: (...args: any[]) => Promise<void>;
+  _changed?: (...args: any[]) => Promise<void>;
+  _movedBefore?: (...args: any[]) => Promise<void>;
+  _removed?: (...args: any[]) => Promise<void>;
```
```suggestion
  _added?: (...args: any[]) => Promise<void> | void;
  _addedBefore?: (...args: any[]) => Promise<void> | void;
  _changed?: (...args: any[]) => Promise<void> | void;
  _movedBefore?: (...args: any[]) => Promise<void> | void;
  _removed?: (...args: any[]) => Promise<void> | void;
```
We could even do something like:

`class ObserveHandle<T = any>` and those methods be something like `(...args: T[]) => Promise<void> | void;`

And in the instance of the handle we could pass in the type of the collection(at least the name idk)

Grubba27 commented on packages/mongo/observe_multiplex.ts:
```diff
@@ -140,7 +140,7 @@ export class ObserveMultiplexer {
   }
 
-  _applyCallback(callbackName: string, args: any[]): void {
+  async _applyCallback(callbackName: string, args: any[]) {
     this._queue.queueTask(async () => {
```
is really necessary turning this function to async? (this is not awaitable–as we do not [return the `scheduleRun` in AsyncQueue](https://github.com/meteor/meteor/blob/d27946d0c0bdc8a9e41dd4b37e828327d480560c/packages/meteor/async_helpers.js#L54-L61)

leonardoventurini replied:
No need to be async, adjusted that

Grubba27 commented on packages/ddp-server/stream_server.js:
```diff
@@ -14,12 +15,18 @@ import once from 'lodash.once';
-        ? JSON.parse(process.env.SERVER_WEBSOCKET_COMPRESSION) : {};
+  var websocketCompressionConfig = process.env.SERVER_WEBSOCKET_COMPRESSION ?
+    JSON.parse(process.env.SERVER_WEBSOCKET_COMPRESSION) : {};
+
   if (websocketCompressionConfig) {
-    extensions.push(Npm.require('permessage-deflate').configure(
-      websocketCompressionConfig
-    ));
+    extensions.push(Npm.require('permessage-deflate2').configure({
+      threshold: 512,
+      level: zlib.constants.Z_BEST_SPEED,
+      memLevel: zlib.constants.Z_MIN_MEMLEVEL,
+      noContextTakeover: true,
+      maxWindowBits: zlib.constants.Z_MIN_WINDOWBITS,
+      ...(websocketCompressionConfig || {})
+    }));
   }
```
Are those configs documented? this whole config could become even a guide for increasing performance

leonardoventurini replied:
I don't think there are docs for it, I can add them

leonardoventurini replied:
Added them https://github.com/meteor/meteor/pull/13445/commits/c73515470f5a4a623b8403231fbfa6c8472cd5b1

Grubba27 commented on packages/ddp-server/package.js:
```diff
@@ -5,12 +5,13 @@ Package.describe({
 Npm.depends({
-  "permessage-deflate": "0.1.7",
+  "permessage-deflate2": "0.1.8-alpha.1",
   sockjs: "0.3.24",
```
How hard would it be to change from sockjs? Do we have testing covering the usage?

leonardoventurini replied:
It should involve some work, socket.io is a strong candidate, however I'm not sure there are direct tests for it

Grubba27 commented on packages/meteor/asynchronous_queue.js:
```diff
@@ -0,0 +1,154 @@
+   * Creates a queue that processes tasks in parallel batches while preserving completion order
+   * when needed. Configurable batch size and concurrency limits help optimize throughput.
+   */
+  constructor({
+                batchSize = 50,
+                maxConcurrent = 10,
+                orderMatters = false
+              } = {}) {
```
All usages of  `Meteor._AsynchronousQueue` are with the default params; maybe those options could become env variables?

leonardoventurini committed:
add summarization command

leonardoventurini committed:
make asynchronous queue name backwards compatible

leonardoventurini committed:
use original logic instead of the one with lru map

leonardoventurini committed:
bump threshold

leonardoventurini committed:
adjust types

leonardoventurini committed:
fix order of operations

leonardoventurini committed:
no need for async here

leonardoventurini committed:
order matters unless told otherwise

leonardoventurini committed:
Merge branch 'release-3.1.1' into leonardo/refactor-and-perf-phase-2

leonardoventurini committed:
add adding tasks to beginning of queue

leonardoventurini committed:
Revert "add adding tasks to beginning of queue"

This reverts commit 93319b9d1125f114bcf972994e37d192725e451a.

leonardoventurini committed:
Refactor AsynchronousQueue constructor to use environment variables for batch size and concurrency limits

- Updated the constructor to read batch size and maximum concurrent tasks from environment variables (METEOR_ASYNC_QUEUE_BATCH_SIZE and METEOR_ASYNC_QUEUE_MAX_CONCURRENT) with default values.
- Improved documentation to clarify configuration options for batch processing and task order preservation.

leonardoventurini committed:
adjust types

leonardoventurini committed:
add compression docs

Grubba27 commented on packages/mongo/observe_handle.ts:
```diff
@@ -4,24 +4,30 @@ let nextObserveHandleId = 1;
+  public initialAddsSent: Promise<void>
 
-  constructor(multiplexer: any, callbacks: Record<ObserveHandleCallback, any>, nonMutatingCallbacks: boolean) {
+  _added?: Callback<T>;
+  _addedBefore?: Callback<T>;
+  _changed?: Callback<T>;
+  _movedBefore?: Callback<T>;
+  _removed?: Callback<T>;
+
```
Should not be `T[]`? These methods can send a list of args to the callback https://docs.meteor.com/api/collections#callbacks-1

nachocodoner reviewed (approved):
Great work 🚀

leonardoventurini committed:
add examples

leonardoventurini committed:
update permessage-deflate2

leonardoventurini committed:
add disclaimer

leonardoventurini committed:
downgrade mongo driver minor due to connection issue with old databases
